backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -0,0 +1,188 @@
+
+        Returns
+        -------
+        cell_counts_df: pandas.Series
+            A series with index "cell_type_ontology_term_id" and values "n_cells".
+            This series is the cell counts dataframe with the following modification:
+            1. The cell type ontology term ids that are not in the input cell counts dataframe are added with 0 counts.
+
+        cell_counts_df_rollup: pandas.Series
+            A series with index "cell_type_ontology_term_id" and values "n_cells".
+            "n_cells" contains the sum of "n_cells" for the cell type and its descendants.
+        """
+
+        cell_counts_df = (
+            cell_counts_df.groupby("cell_type_ontology_term_id").sum(numeric_only=True)[["n_cells"]].reset_index()
+        )
+
+        to_attach = pd.DataFrame()
+        to_attach["cell_type_ontology_term_id"] = [
+            i for i in self.all_cell_type_ids if i not in cell_counts_df["cell_type_ontology_term_id"].values
+        ]
+        to_attach["n_cells"] = 0
+        cell_counts_df = pd.concat([cell_counts_df, to_attach], axis=0)
+        cell_counts_df_rollup = rollup_across_cell_type_descendants(cell_counts_df)
+        cell_counts_df = cell_counts_df.set_index("cell_type_ontology_term_id")["n_cells"]
+        cell_counts_df_rollup = cell_counts_df_rollup.set_index("cell_type_ontology_term_id")["n_cells"]
+        return cell_counts_df, cell_counts_df_rollup
+
+    def _delete_unknown_terms_from_ontology_graph(self, node):
+        """
+        Deletes nodes that have no name in the CL ontology from the ontology graph.
+
+        Arguments
+        ---------
+        node: dict
+            A node of the ontology graph. At the top-level, this should be the root node of the ontology graph.
+        """
+        new_children = []
+        for child in node.get("children", []):
+            unknown = child["name"].startswith("CL:")
+            if not unknown:
+                new_children.append(child)
+        if len(new_children) > 0:
+            node["children"] = new_children
+        elif "children" in node:
+            del node["children"]
+
+        for child in node.get("children", []):
+            self._delete_unknown_terms_from_ontology_graph(child)
+
+    def _initialize_id_to_name(self):
+        """
+        Initializes a dictionary that maps cell type ontology term ids to their names.
+
+        Returns
+        -------
+        id_to_name: dict
+            A dictionary that maps cell type ontology term ids to their names.
+        all_cell_type_ids: list
+            A list of all cell type ontology term ids.
+        """
+        id_to_name = {}
+        for term in self.ontology.terms():
+            if term.id.startswith("CL:"):
+                id_to_name[term.id] = term.name
+        all_cell_type_ids = list(id_to_name.keys())
+        return id_to_name, all_cell_type_ids
+
+    def _traverse_ontology_with_counting(self, node, traverse_node_counter=None, all_unique_nodes=None):
```

Comment:
also an actual code comment on this line: you can add an actual type annotation to the parameter, like:
```
def _traverse_ontology_with_counting(self, node, traverse_node_counter: Optional[dict[str, Any]], all_unique_nodes: Optional[set[str]]):
```
which avoids the need to document the type in the code comments and is more useful since you can use the type annotations to enforce the caller to pass in the right type